### PR TITLE
inspection: types nullable

### PR DESCRIPTION
**Pull request by Hidanio**

### CODE REVIEW

File: src/linter/root_checker.go [hunk 1]
```diff
@@ -604,8 +604,20 @@ func (r *rootChecker) checkFuncParam(p *ir.Parameter) {
 		}
 		return true
 	})
-
 	r.CheckTypeHintFunctionParam(p)
+	r.CheckParamNullability(p)
+}
+
+func (r *rootChecker) CheckParamNullability(p *ir.Parameter) {
+	if param, ok := p.VariableType.(*ir.Name); ok {
```
[the remark this answers is not in the document]
Comment (Hidanio) [reply src/linter/root_checker.go]: ok

File: src/tests/golden/testdata/quickfix/stringNullable.fix.expected
Comment (Danil42Russia) [inline src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: are you sure you need this file?

File: src/tests/golden/testdata/quickfix/nullableTypes.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+class MyClass2 {
+    public function myMethod(MyClass1 $a = null) {
+        return 0;
+    }
+}
+
+function nullableArray(array $a = null) {
+	return 0;
+}
```
Comment (Danil42Russia) [inline src/tests/golden/testdata/quickfix/nullableTypes.php:22]: line break

File: src/tests/golden/testdata/quickfix/nullableTypes.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+function f(string $filed = null) {
```
Comment (Danil42Russia) [inline src/tests/golden/testdata/quickfix/nullableTypes.php:3]: let's rename the function

File: src/tests/golden/testdata/quickfix/nullableTypes.php
Comment (Danil42Russia) [inline src/tests/golden/testdata/quickfix/nullableTypes.php:1]: let's add a test when there are multiple parameters

File: src/linter/report.go [hunk 4]
```diff
@@ -25,6 +25,15 @@ func addBuiltinCheckers(reg *CheckersRegistry) {
 			After:    `$s = strip_tags($s, '<br>')`,
 		},
 
+		{
+			Name:     "nullableType",
+			Default:  true,
+			Quickfix: true,
+			Comment:  "Report not nullable string can be null.",
+			Before:   `function f(string $str = null);`,
```
Comment (Danil42Russia) [inline src/linter/report.go:33]: Let's change the wording. It's not just `string`.

File: src/linter/quickfix.go [hunk 5]
```diff
@@ -47,6 +48,34 @@ func (g *QuickFixGenerator) NullForNotNullableProperty(prop *ir.PropertyStmt) qu
...
+		}
+		value = v.Value
+	case *ir.Identifier:
+		pos = &position.Position{
+			StartPos: v.Position.StartPos,
+			EndPos:   v.Position.EndPos,
+		}
+		value = v.Value
+	default:
+		panic("unexpected type")
```
Comment (Danil42Russia) [inline src/linter/quickfix.go:69]: let's not panic

File: src/tests/golden/testdata/quickfix/stringNullable.fix.expected
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: Yes

File: src/tests/golden/testdata/quickfix/nullableTypes.php
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php:1]: Ok

File: src/tests/golden/testdata/quickfix/nullableTypes.php (hunk 3 above)
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php]: Done

File: src/tests/golden/testdata/quickfix/nullableTypes.php (hunk 2 above)
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php]: Done

File: src/linter/report.go (hunk 4 above)
Comment (Hidanio) [reply src/linter/report.go:33]: Done

File: src/linter/quickfix.go (hunk 5 above)
Comment (Hidanio) [reply src/linter/quickfix.go]: Removed

File: src/tests/golden/testdata/quickfix/stringNullable.fix.expected
Comment (Danil42Russia) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: why?

Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: You can check `quickfix_test.go`. We have test.php & test.php.fix.expected; In proccess will generate test.php.fix and compare with test.php.fix.expected, if everything ok - test.php.fix  will be delete, otherwise - no

Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: `otherwise - no` i mean it will help us to debug (view generated fix)

Comment (Danil42Russia) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: 1) You don't have `stringNullable` in your project
2) The file `stringNullable.fix.expected`, does not have the extension `.php`
3) I deleted the file, nothing is broken

File: src/tests/golden/testdata/quickfix/nullableTypes.php
Comment (Danil42Russia) [reply src/tests/golden/testdata/quickfix/nullableTypes.php:1]: where?

Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php:1]: `nullable_types_test.go`

File: src/tests/golden/testdata/quickfix/stringNullable.fix.expected
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/stringNullable.fix.expected:1]: Refresh git. I think you missed something

File: src/tests/golden/testdata/quickfix/nullableTypes.php
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php:1]: Added for quickfix too

File: src/linter/root_checker.go [hunk 6]
```diff
@@ -604,8 +604,32 @@ func (r *rootChecker) checkFuncParam(p *ir.Parameter) {
...
+		}
+		paramType = paramIdentifier
+	}
+
+	defValue, defValueOk := p.DefaultValue.(*ir.ConstFetchExpr)
+	if !defValueOk {
+		return
+	}
+
+	if defValue.Constant.Value != "null" {
```
Comment (Danil42Russia) [inline src/linter/root_checker.go:627]: `mixed` is an independent type. It already includes `null`
We need to ignore this case.

File: src/tests/golden/testdata/quickfix/nullableTypes.php.fix.expected [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    public function myMethod(?MyClass1 $a = null) {
+        return 0;
+    }
+}
+
+function nullableArray(?array $a = null) {
+	return 0;
+}
+
+function multipleArgsExample(string $a, ?int $b = null, ?bool $c = null) {
```
Comment (Danil42Russia) [inline src/tests/golden/testdata/quickfix/nullableTypes.php.fix.expected:24]: Add a test for `null|T`

File: src/linter/root_checker.go (hunk 6 above)
Comment (Hidanio) [reply src/linter/root_checker.go:631]: Added

File: src/tests/golden/testdata/quickfix/nullableTypes.php.fix.expected (hunk 7 above)
Comment (Hidanio) [reply src/tests/golden/testdata/quickfix/nullableTypes.php.fix.expected]: Added

**Review by Danil42Russia — APPROVED** [review]

File: src/tests/checkers/nullable_types_test.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+func TestNotNullableMultipleArgs(t *testing.T) {
+	test := linttest.NewSuite(t)
+	test.AddFile(`<?php
+function multipleArgsExample(string $a, int $b = null, bool $c = null) {
+	return 0;
+}
+`)
+
+	test.Expect = []string{
```
[the remark this answers is not in the document]
Comment (Hidanio) [reply src/tests/checkers/nullable_types_test.go]: `string $a` have no null

File: src/linter/report.go [hunk 9]
```diff
@@ -25,6 +25,15 @@ func addBuiltinCheckers(reg *CheckersRegistry) {
 			After:    `$s = strip_tags($s, '<br>')`,
 		},
 
+		{
+			Name:     "notExplicitNullableParam",
+			Default:  true,
+			Quickfix: true,
+			Comment:  "Report not nullable param can be null.",
```
Comment (i582) [inline src/linter/report.go:32]: Maybe
> Report not nullable param with explicit `null` default value

?

**Comment by i582** [comment]
🎉
